awslabs/aws-lambda-powertools-typescript
# feat(tracer): Add capability to continue traces comming from SQS triggers

## 🚧 WIP 🚧  :
* Missing proper unit and e2e tests assertions
* Need to evaluate the API design : 
  * is continueSQSRecordTrace and the way to invoke it good enough ? 
  * Should we store toCloseSegments on our side and provide a function to close them ?
* can we really can't do better in term of implementation ??? 
  * might worth to contact a xray sdk node developer 
  * check in deeper details what is done here https://github.com/Sam-Martin/aws-xray-lambda-segment-shim


## Description of your changes


This PR is a first draft trying to give a solution to the Xray node sdk issue https://github.com/aws/aws-xray-sdk-node/issues/208 where traces should be continued when lambdaare triggered from SQS: `Producer -> SQS -> Consumer Lambda -> ...` (Today it creates 2 traces: `Producer -> SQS` and `Lambda -> ...`).

Here are the findings while implementing this:

* aws-xray-sdk-node has a specific behavior for lambda based on `process.env.LAMBDA_TASK_ROOT` (https://github.com/aws/aws-xray-sdk-node/blob/master/packages/core/lib/aws-xray.js#L381)
* Even if active tracing is disabled on the consumer lambda a trace is created for the Consumer Lambda
* If XRay permission is not granted to consumer then no traces are created.
* We can force env var _X_AMZN_TRACE_ID but 
  * need to be set before import of the xray-sdk-core
  * it does not prevent the extra trace to be created with a different traceID
  * It does not help with the multi message use case

### How to verify this change

```
cd packages/tracing
export DISABLE_TEARDOWN=true
npx jest --group=e2e/tracer/sqs
```

Check Xray for traces
![Screenshot 2021-12-29 at 10 18 44](https://user-images.githubusercontent.com/1225809/147646926-e95c17aa-9f35-49b3-801a-12869220deb9.png)

Or Cloudwatch Service Map
![Screenshot 2021-12-29 at 10 20 07](https://user-images.githubusercontent.com/1225809/147646947-7150fd1c-8e6e-4a2c-ac0d-49654f88e8e8.png)
![Screenshot 2021-12-29 at 10 20 24](https://user-images.githubusercontent.com/1225809/147646943-a7f6b6ef-e7ef-47e0-ae2c-397f3bfad8ee.png)





### Related issues, RFCs

#286

### PR status

***Is this ready for review?:*** NO  
***Is it a breaking change?:*** NO

## Checklist


### Breaking change checklist


---

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.
